crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -232,44 +241,94 @@ impl CumulocityConverter {
+    fn auto_register_entity(&mut self, entity_id: &str) -> Vec<Message> {
+        let mut register_messages = vec![];
+        let (device_id, service_id) = match entity_id.split('/').collect::<Vec<&str>>()[..] {
+            ["device", device_id, "service", service_id, ..] => (device_id, Some(service_id)),
+            ["device", device_id, "", ""] => (device_id, None),
+            _ => return register_messages,
+        };
+
+        // register device if not registered
+        let device_topic = format!("{MQTT_ROOT}/device/{device_id}//");
+        if self.entity_store.get(&device_topic).is_none() {
+            let device_type = if device_id == "main" {
+                "device"
+            } else {
+                "child-device"
+            };
+            let device_name = if device_id == "main" {
+                self.entity_store.main_device_name()
+            } else {
+                device_id
+            };
+            let device_register_payload =
+                format!("{{ \"@type\":\"{device_type}\", \"@id\":\"{device_name}\"}}");
+            let device_register_message =
+                Message::new(&Topic::new(&device_topic).unwrap(), device_register_payload)
+                    .with_retain();
+            register_messages.push(device_register_message.clone());
+            self.entity_store
+                .update(device_register_message.try_into().unwrap())
+                .unwrap();
+        }
+
+        // register service itself
+        if let Some(service_id) = service_id {
+            let service_topic = format!("{MQTT_ROOT}/device/{device_id}/service/{service_id}");
+            let service_register_payload = r#"{"@type": "service", "type": "systemd"}"#.to_string();
+            let service_register_message = Message::new(
+                &Topic::new(&service_topic).unwrap(),
+                service_register_payload,
+            )
+            .with_retain();
+            register_messages.push(service_register_message.clone());
+            self.entity_store
+                .update(service_register_message.try_into().unwrap())
+                .unwrap();
+        }
+
+        register_messages
+    }
+
     fn try_convert_measurement(
         &mut self,
         input: &Message,
     ) -> Result<Vec<Message>, ConversionError> {
         let mut mqtt_messages: Vec<Message> = Vec::new();
 
-        let maybe_child_id = get_child_id_from_measurement_topic(&input.topic.name);
-        let c8y_json_payload = match maybe_child_id {
-            Some(child_id) => {
-                // Need to check if the input Thin Edge JSON is valid before adding a child ID to list
-                let c8y_json_child_payload =
-                    json::from_thin_edge_json_with_child(input.payload_str()?, child_id.as_str())?;
-                add_external_device_registration_message(
-                    child_id,
-                    &mut self.children,
-                    &mut mqtt_messages,
-                );
```

Comment:
This has been moved to the `try_convert()` method https://github.com/thin-edge/thin-edge.io/pull/2169/files#diff-1d1eb076015e0658044f7b03b0b805fce4f0253f4048587797fc5dd42019f908L149.

- Pros: The registration can then be done for all kind of messages (measurements/alarms/events)
- Cons: Now an ill-formed message trigger the auto-registration while it should not.
- Because of that the test `convert_first_measurement_invalid_then_valid_with_child_id` must be ignored https://github.com/thin-edge/thin-edge.io/pull/2169/files#diff-08a20a8877f5e1e1bce1aac2c6fc0e1a63a36cc9311e7e60cb3aacacf3c8c57cR1361